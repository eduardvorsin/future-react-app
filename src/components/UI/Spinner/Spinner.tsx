import React from 'react';
import { useTranslation } from 'react-i18next';
import classes from './Spinner.module.css';

type SpinnerProps = {
  className?: string,
  size: number,
};

export interface SpinnerStyles extends React.CSSProperties {
  '--size': string;
}

export interface RingStyles extends React.CSSProperties {
  '--ring-size': string;
}

const Spinner: React.FC<SpinnerProps> = ({
  className,
  size = 60,
}) => {
  const { t } = useTranslation();
  const spinnerClasses = [
    classes.spinner,
    className,
  ].join(' ');

  const spinnerStyles: SpinnerStyles = {
    '--size': `${size}px`,
  };

  const ringStyles: RingStyles = {
    '--ring-size': `${0.8 * size}px`,
  };

  return (
    <div
      className={spinnerClasses}
      aria-label={`${t('loading')}...`}
      style={spinnerStyles}
      title="loading spinner"
    >
      <div
        className={classes.spinner__ring}
        style={ringStyles}
      />
      <div
        className={classes.spinner__ring}
        style={ringStyles}
      />
      <div
        className={classes.spinner__ring}
        style={ringStyles}
      />
      <div
        className={classes.spinner__ring}
        style={ringStyles}
      />
    </div>
  );
};

export default Spinner;
