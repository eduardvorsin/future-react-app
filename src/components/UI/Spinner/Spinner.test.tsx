/* eslint-disable @typescript-eslint/no-empty-function */
import React from 'react';
import { render, screen } from '@testing-library/react';
import { I18nextProvider } from 'react-i18next';
import Spinner from './Spinner';
import i18n from '../../../localization/i18next';

describe('Spinner tests', () => {
  beforeAll(() => {
    jest.mock('react-i18next', () => ({
      ...jest.requireActual('react-i18next'),
      useTranslation: () => ({
        t: (str: string) => str,
        i18n: {
          changeLanguage: () => new Promise(() => { }),
        },
      }),
    }));
  });

  it('is rendered correctly', () => {
    render(
      <Spinner
        size={100}
      />,
    );

    expect(screen.getByTitle<HTMLDivElement>('loading spinner')).toBeInTheDocument();
  });

  it('if the prop size is 120 then the width and height of the spinner is also 120', () => {
    render(
      <Spinner
        size={120}
      />,
    );

    expect(screen.getByTitle<HTMLDivElement>('loading spinner')).toHaveStyle({
      width: '120',
      height: '120',
    });
  });

  it('is a basic snapshot', () => {
    render(
      <Spinner
        size={120}
      />,
    );

    expect(screen.getByTitle<HTMLDivElement>('loading spinner')).toMatchSnapshot();
  });

  it('is a snapshot with a component size of 150px', () => {
    render(
      <Spinner
        size={150}
      />,
    );

    expect(screen.getByTitle<HTMLDivElement>('loading spinner')).toMatchSnapshot();
  });
});

describe('Spinner integration tests', () => {
  it('the component with the text in English is displayed correctly', () => {
    const localizedText = `${i18n.t('loading')}...`;

    render(
      <I18nextProvider i18n={i18n}>
        <Spinner
          size={100}
        />
      </I18nextProvider>,
    );

    expect(screen.getByLabelText<HTMLDivElement>(localizedText)).toBeInTheDocument();
  });

  it('the component with the text in Russian is displayed correctly', () => {
    i18n.changeLanguage('ru');
    const localizedText = `${i18n.t('loading')}...`;

    render(
      <I18nextProvider i18n={i18n}>
        <Spinner
          size={100}
        />
      </I18nextProvider>,
    );

    expect(screen.getByLabelText<HTMLDivElement>(localizedText)).toBeInTheDocument();
  });
});
