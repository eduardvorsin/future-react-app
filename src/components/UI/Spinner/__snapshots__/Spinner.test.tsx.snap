// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spinner tests is a basic snapshot 1`] = `
<div
  aria-label="Loading..."
  class="spinner "
  style="--size: 120px;"
  title="loading spinner"
>
  <div
    class="spinner__ring"
    style="--ring-size: 96px;"
  />
  <div
    class="spinner__ring"
    style="--ring-size: 96px;"
  />
  <div
    class="spinner__ring"
    style="--ring-size: 96px;"
  />
  <div
    class="spinner__ring"
    style="--ring-size: 96px;"
  />
</div>
`;

exports[`Spinner tests is a snapshot with a component size of 150px 1`] = `
<div
  aria-label="Loading..."
  class="spinner "
  style="--size: 150px;"
  title="loading spinner"
>
  <div
    class="spinner__ring"
    style="--ring-size: 120px;"
  />
  <div
    class="spinner__ring"
    style="--ring-size: 120px;"
  />
  <div
    class="spinner__ring"
    style="--ring-size: 120px;"
  />
  <div
    class="spinner__ring"
    style="--ring-size: 120px;"
  />
</div>
`;
